src/history/api.rs

```diff
@@ -1,152 +1,348 @@
+use super::model::{Blob, Contract, Proof, Transaction};
 use crate::{
-    history::{BlocksFilter, ContractsFilter, TransactionsFilter},
-    model::BlockHeight,
+    model::{Block, BlockHeight},
     rest::RouterState,
 };
 use axum::{
     extract::{Path, Query, State},
     http::StatusCode,
-    response::IntoResponse,
     Json,
 };
+use serde::{de::DeserializeOwned, Deserialize};
+use tracing::error;
 
-pub async fn get_transaction(
-    Path(tx_hash): Path<String>,
-    State(state): State<RouterState>,
-) -> Result<impl IntoResponse, StatusCode> {
-    match state.history.lock().await.transactions.get(&tx_hash) {
-        Ok(Some(tx)) => Ok(Json(tx)),
-        Ok(None) => Err(StatusCode::NOT_FOUND),
-        Err(_) => Err(StatusCode::INTERNAL_SERVER_ERROR),
+#[derive(Deserialize, Debug)]
+pub struct Filters {
+    pub skip: Option<usize>,
+    pub reverse: Option<bool>,
+    pub limit: Option<usize>,
+}
+
+fn filter_iter<T: DeserializeOwned>(iter: crate::history::db::Iter<T>, filters: Filters) -> Vec<T> {
+    // from the start
+    if !filters.reverse.unwrap_or(false) {
+        iter.rev() // just for this !
+            .skip(filters.skip.unwrap_or(0))
+            .take(filters.limit.unwrap_or(10))
+            .flat_map(|i| {
+                i.map(|i| i.value().map_err(|e| error!("deserializing  data: {}", e)))
+                    .map_err(|e| error!("iterating over data: {}", e))
+            })
+            .filter_map(|i| i.ok())
+            .collect::<Vec<T>>()
+    } else {
+        iter.skip(filters.skip.unwrap_or(0))
+            .take(filters.limit.unwrap_or(10))
+            .flat_map(|i| {
+                i.map(|i| i.value().map_err(|e| error!("deserializing  data: {}", e)))
+                    .map_err(|e| error!("iterating over data: {}", e))
+            })
+            .filter_map(|i| i.ok())
+            .collect::<Vec<T>>()
```

Comment:
Suggestion (je pense que ça marche)
```suggestion
    let iter = if !filters.reverse.unwrap_or(false) {
        iter.rev()
    } else {
        iter
    }
    iter.skip(filters.skip.unwrap_or(0))
        .take(filters.limit.unwrap_or(10))
        .flat_map(|i| {
            i.map(|i| i.value().map_err(|e| error!("deserializing  data: {}", e)))
                .map_err(|e| error!("iterating over data: {}", e))
        })
        .filter_map(|i| i.ok())
        .collect::<Vec<T>>()
```

Reply:
nope it doesn't work. the then and else branches do not have the same type (I already tried this).